### PR TITLE
Add RectorConfig for custom configuration methods and avoid conflicts with Symfony API

### PR DESCRIPTION
**Pull request by TomasVotruba**
Follow up to https://github.com/rectorphp/rector-src/pull/1891

---

Using Symfony container for `rector.php` configuration is handy for autocomplete, static analysis and re-use of Symfony DI container.
Yet, it has few drawbacks, when it comes to a project that runs on Symfony with it's own Symfony DI. There 2 classes of same name can conflict in IDE and come to an autoloading issue. See https://github.com/rectorphp/rector/discussions/7035 for more.

@Wirone came with a great idea, to have an own configurator class, that will be tailored to Rector's needs and features.
Similar to Config Builder Classes from Symfony: https://symfony.com/blog/new-in-symfony-5-3-config-builder-classes

## How do we use `rector.php` now?

```php
use Symfony\Component\DependencyInjection\Loader\Configurator\ContainerConfigurator;
use Rector\Core\Configuration\Option;
use Rector\DowngradePhp81\Rector\Property\DowngradeReadonlyPropertyRector;

// ups, possible conflict with ContainerConfigurator
return static function (ContainerConfigurator $containerConfigurator): void {
    $parameters = $containerConfigurator->parameters();

    // too verbose params, constants and possible typo in param value 
    $parameters->set(Option::PATHS, [[ // ups, "[[" typo
        __DIR__.'/src/',
    ]]);

    $services = $containerConfigurator->services();
    $services->set(DowngradeReadonlyPropertyRector::class);
};
```

<br>

## How could we use `RectorConfig` in future?

```php
use Rector\Config\RectorConfig;

return static function (RectorConfig $rectorConfig): void {
    $rectorConfig->parallel();

    $rectorConfig->paths([
        __DIR__.'/src',
    ]);


    $rectorConfig->rule(DowngradeReadonlyPropertyRector::class);
};

```

### CODE REVIEW

**Comment by samsonasik** [comment]
new e2e directory need to be register to e2e.yaml

https://github.com/rectorphp/rector-src/blob/0a6b3b3ad45a8f128cee78c69f789c47c427d367/.github/workflows/e2e.yaml#L25-L32

**Comment by samsonasik** [comment]
custom config may need to be conditionally checked

https://github.com/rectorphp/rector-src/blob/0a6b3b3ad45a8f128cee78c69f789c47c427d367/.github/workflows/e2e.yaml#L53-L55

**Comment by TomasVotruba** [comment]
> new e2e directory need to be register to e2e.yaml

Hm :thinking: We should automate this with some script.

**Comment by samsonasik** [comment]
custom config check need to be checked as well

https://github.com/rectorphp/rector-src/blob/0a6b3b3ad45a8f128cee78c69f789c47c427d367/.github/workflows/e2e.yaml#L57-L59

with `||` condition, ref

https://github.community/t/and-operator-in-if-condition/154825

```
if: ${{ (matrix.directory == 'e2e/parallel-custom-config') || (matrix.directory == 'e2e/use-rector-configurator')}}
```

also in line 64 with `&&` condition`

```
if: ${{ (matrix.directory != 'e2e/parallel-custom-config') &&  (matrix.directory != 'e2e/use-rector-configurator') }}
```

**Comment by TomasVotruba** [comment]
@samsonasik > custom config check need to be checked as well

I see. I've just copy-pasted another directory, this can be in the root. I've moved it to avoid this one. Thanks

File: e2e/use-rector-configurator/src/SomeClass.php [hunk 1]
```diff
@@ -0,0 +1,14 @@
+<?php
+
+declare(strict_types=1);
+
+namespace E2e\Parallel\Custom\Config;
```
Comment (samsonasik) [inline e2e/use-rector-configurator/src/SomeClass.php]: namespace may need update based on e2e dir,

Comment (TomasVotruba) [reply e2e/use-rector-configurator/src/SomeClass.php]: :+1:

File: composer.json [hunk 2]
```diff
@@ -153,7 +153,8 @@
         "enable-patching": true,
         "patches": {
             "symfony/dependency-injection": [
-                "https://raw.githubusercontent.com/rectorphp/vendor-patches/main/patches/symfony-dependency-injection.patch"
+                "https://raw.githubusercontent.com/rectorphp/vendor-patches/main/patches/symfony-dependency-injection.patch",
+                "https://raw.githubusercontent.com/rectorphp/vendor-patches/main/patches/symfony-php-config-loader.patch"
             ]
```
Comment (TomasVotruba) [inline composer.json:158]: After this PR is merged, line needs to update in each Rector split repository's `composer.json`, so we have the same patched Symfony everywhere

**Comment by samsonasik** [comment]
target-repository e2e may need update, for example on global install e2e https://github.com/rectorphp/rector-src/tree/main/build/target-repository/e2e/global-install

File: packages/Config/RectorConfig.php [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+
+/**
+ * @api
+ * Same as Symfony container configurator, with patched return type for "set()" method for easier DX.
+ * It is an alias for internal class that is prefixed during build, so it's basically for keeping stable public API.
+ */
+final class RectorConfig extends ContainerConfigurator
+{
+    /**
+     * @param string[] $paths
```
Comment (samsonasik) [inline packages/Config/RectorConfig.php]:
```
@param mixed[] $paths
```

may be used when next `Assert::allString($paths)` is checked.

**Comment by TomasVotruba** [comment]
Let's :ship: it :+1:

File: packages/Config/RectorConfig.php (hunk 3 above)
Comment (TomasVotruba) [reply packages/Config/RectorConfig.php]: :+1:

**Pull request merged**